File: 프로그래머스/lv3/49191. 순위/순위.js
function solution(n, results) {
  var answer = 0;

  const g = Array.from({ length: n }, (_, i) => {
    return Array.from({ length: n }, (_, j) => (i == j ? 0 : false));
  });

  results.forEach(([a, b]) => {
    g[a - 1][b - 1] = 1;
    g[b - 1][a - 1] = -1;
  });

  for (let m = 0; m < n; m++) {
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        if (g[i][m] == 1 && g[m][j] == 1) {
          g[i][j] = 1;
        }
        if (g[i][m] == -1 && g[m][j] == -1) {
          g[i][j] = -1;
        }
      }
    }
  }

  for (let i = 0; i < n; i++) {
    let flag = false;
    for (let j = 0; j < n; j++) {
      if (g[i][j] === false) {
        flag = true;
        break;
      }
    }
    answer = !flag ? answer + 1 : answer;
  }

  return answer;
}

File: 프로그래머스/3/49191. 순위/순위.py
# 플로이드 와샬 알고리즘을 유사하게 활용한다.
# 이 때 최단 거리까지 계산할 필요는 없고, 
# 거쳐가는 경로(k)가 있을 때 i -> j에 1을 넣어준다.
def solution(n, results):
    answer = 0
    graph = [[0 for _ in range(n)] for _ in range(n)]
    for result in results:
        winner = result[0] - 1
        loser = result[1] - 1
        graph[winner][loser] = 1
        
    # k = 거쳐가는 노드
    for k in range(n):
        # i = 출발 노드
        for i in range(n):
            # j = 도착 노드
            for j in range(n):
                if not graph[i][j]:
                    if graph[i][k] and graph[k][j]:
                        graph[i][j] = 1
    rank_determined = [0 for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if graph[i][j] == 1:
                rank_determined[i] += 1
                rank_determined[j] += 1
    for rank in rank_determined:
        if rank == n - 1:
            answer += 1
    
    return answer